### PR TITLE
audio avoid freeze on background 

### PR DESCRIPTION
**Pull request by simon300000**
#65

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -1953,6 +1966,17 @@ ehDownloadAction.innerHTML = ehDownloadArrow + ' <a>Download Archive</a>';
...
 
+	if (setting['play-silent-music']) {
+    oscillator = audioCtx.createOscillator()
+    gainNode = audioCtx.createGain()
+    oscillator.connect(gainNode)
+    gainNode.connect(audioCtx.destination)
+    oscillator.frequency.value = 1
+    gainNode.gain.value = 0.001
+    oscillator.start()
+	}
```
Comment (ccloli) [inline src/main.js]: Please use tab to indent instead of 2-spaces, also add a semicolon (`;`) after each line.

Comment (simon300000) [inline src/main.js]: sorry 坏习惯

**Comment by ccloli** [comment]
Here is a merge conflict after merging another PR. If you want to fixed it, use `git rebase master` from your side, then solve the conflict in each rebase error. If you are not sure how to do it, here is the steps:

First conflict should be in 9f4d3ce (_Control silent music on advance setting_), and you **only** needs to edit `main.js`. Please use the changes what you did in your master branch (replace symbols to numbers), but also keeps your added new setting, just replace `++` to the number `(4)` (remember to replace the original `(4)` to `(5)`). Then run `build.bat` or your `build.sh` to overwrite the `user.js` file (also fixed the conflict in that `user.js` file). Use `git add .` and `git rebase --continue`.

The next conflict should be in 3bd8d08 (_Misspell_), just do what you did again, then build, `git add .`, `git rebase --continue` again.

Then there should be no more conflict, but if you have conflict in your new commit 9534d73 (_Update from master_), just use `git rebase --skip` since you have updated them in previous steps.

Then you can use `git push` to push to GitHub (you may need `git push --force` if remote rejects your request).

If you are still not sure how to do it, leave a comment here and I'll do the job.

**Comment by simon300000** [comment]
There seems having some problem

**Comment by ccloli** [comment]
@simon300000 no need to close, fix it with my steps, then push it, that should fix the problem.

**Comment by simon300000** [comment]
Thank you, it seems good now

**Comment by simon300000** [comment]
If this still not help with the problem, it might need to stop the music and start again in each half second, this is what I done in the testing script

**Comment by ccloli** [comment]
LGTM, is there anything going to be added?

**Comment by simon300000** [comment]
Nothing need to do, if that work

**Comment by ccloli** [comment]
Got an error when downloading 5 archives at the same time:
`ERROR: Execution of script 'E-Hentai Downloader' failed! Failed to construct 'AudioContext': The number of hardware contexts provided (6) is greater than or equal to the maximum bound (6).`

Looks like it's a hardware or API limit, and seems cannot be fixed.

**Comment by simon300000** [comment]
Ohh wow
Is that happening also if setting is disabled?
simon3000
simon3000@163.com

**Comment by simon300000** [comment]
Hello Do you think this might help:
if the url have some kind of "download=true"
than the script will automatically start download

and user can put the link they want to download in somewhere like localStorage, and the script will jump to the next link if the current one finish with "download=true", so user could download multiple link in one tab?

**Comment by ccloli** [comment]
No, it won't happen if it's disabled or not downloading, but it will throw the error if another 4 (or maybe 5 or 6) tabs are downloading (showing the audio icon). 
```
var audioCtx = !setting['play-silent-music'] || new(unsafeWindow.AudioContext || unsafeWindow.webkitAudioContext)
```
If it reach the limit, opening new tab will throw the error, and the action box will disappeared.

***

Maybe consider it in v2, but at this time, I think I won't add new functions that may refactors the code... And here is a historical question: sometimes it may have errors when saving to harddisk and cannot be caught like Google Chrome's exceeding Blob Storage, so it's not wise to clear the previous downloading data if not sure it's really saved successfully. That's the reason why it doesn't support closing the tab automatically when it finish its work (#25).

**Comment by simon300000** [comment]
I'll change it later let it new AudioContext when download actually start instead page is load
it might help a little bit If user did not start the download

**Comment by ccloli** [comment]
Yes but I think it doesn't fix the problem, maybe try catch will fix the bug, but the new tabs are still not using the feature (thought I think no one will download so many archives at the same time).

Shall we look around for another solution? I found that hook the `setTimeout()` function and run `setTimeout()` in a Web Worker would probably be a solution, but not sure will it influence Tampermonkey itself?

**Comment by simon300000** [comment]
I think it will still throw error but action box might alive

**Comment by simon300000** [comment]
You mean using setTimeout on worker so it need communicate between so avoid freeze
it seems good

**Comment by ccloli** [comment]
Tried overriding `setTimeout` in a Web Worker, it really works in the test. However, the script is not working, ~~looks like in Tampermonkey's sandbox, it cached that `setTimeout`~~. From my test, it downloads 3 or 4 images, then it freezes, switch back to the tab, the page is freezed and more than 4 images say it finished in about 1 second.

image

Tested again, from `console.log()`, it's sure that `setTimeout` and `clearTimeout` has already been override, and the `setTimeout` in worker is not in Tampermonkey's sandbox. So maybe `setTimeout` is not the question but some other functions delay / freeze the process. I've push the code to a new branch `web-worker-timer`, you can have a look.

BTW, `setTimeout` or something else is not the topic of `AudioContext`, maybe we should continue the other discussions in #65.

**Comment by ccloli** [comment]
I replaced `AudioContext()` with a static audio file and `Audio()`, please check if it works. :-)

**Pull request merged**